### ISSUE TITLE
datastore: support Put and Get for JSON entities 

### ISSUE THREAD

**Issue by willie** [GoogleCloudPlatform/google-cloud-go#223]
tl;dr: I'd like to be able to pass a JSON formatted string or the interface value of the marshaled JSON to Put and be able to retrieve it via Get.

Here's my use case:
1. You have a system that generates and emits JSON messages (could be a logger, could be something else). The JSON may have extra data in it from time to time. 
2. These JSON messages arrive via Pub/Sub.
3. One subscriber to these messages is a golang program that does some analysis on the JSON and stores the JSON it receives into Cloud Datastore. The issue is that the Put interface requires that the data stored be a struct (which could result in data loss if the struct doesn't have all the JSON fields) or requires that one implements PropertyListSaver, but that requires a type to be able to implement that interface. Being able to Put/Get JSON directly would be lovely.

**Comment by okdave** [comment]
Hi @willie – I'd like to know more about your use case, cause I'm not sure I completely understand it yet.

When the JSON data arrives, what format do you have it in? Is it still marshalled as `[]byte`, or has it been parsed into a struct or map[string]interface{} or ...?

Do you want to be able to index the data / what type of queries do you need to be able to run?

**Comment by willie** [comment]
It can be either format. Right now, my code is fine with dealing with it as a struct or as a map[string]interface{}, but storing it in a struct to put into datastore may cause data loss.

Short version: JSON events come in via Pub/Sub. Those events have data that I need to take action on based on some of the field values. I will want to store some of those events for later, for long term knowledge. I may not be aware of all the fields that may be coming through at a given time due to software revisions, but I will want them later (they might be queried, they might not).

**Comment by okdave** [comment]
There's no immediate plan to any specific JSON support to the datastore API, but you should be able to achieve what you want using the existing API.

The simplest way is actually to store the marshalled JSON as a `[]byte` in datastore. If you need to index any of the properties, you can do that by storing them as separate properties. You can do this by implementing `datastore.PropertyLoadSaver`:

```go
type Message struct {
  ID      string
  Updated time.Time
  Body    map[string]interface{} `datastore:"-"`
}

func (m *Message) Save() ([]datastore.Property, error) {
  props, err := datastore.SaveStruct(m)
  if err != nil {
    return nil, err
  }
  body, err := json.Marshal(m.Body)
  if err != nil {
    return nil, err
  }
  props = append(props, datastore.Property{
    Name:    "Body",
    Value:   body,
    NoIndex: true,
  })
  return props, nil
}

func (m *Message) Load([]datastore.Property) error { etc. }
```

If you need to index arbitrary fields from the incoming JSON, you should be able to do something similar by defining a `type Message map[string]interface{}` which implements `PropertyLoadSaver`, though that is going to be significantly more involved.

Does that give you sufficient ability to implement what you need?

**Comment by willie** [comment]
Yes, once I figured out how to wrap my fake struct with datastore.PropertyLoadSaver. I appreciate the help, but I think better documentation on using Put/Get and PropertyLoadSaver would be helpful.

**Comment by rumsrami** [comment]
Can you please provide an example on how to Put/Get together with PropertyLoadSaver?

**Comment by kareemgithubprotonme** [comment]
Can you please provide an example on how to Put/Get together with PropertyLoadSaver?
my usecase is that i'm trying to store an indexed json, i can't do that directly using struct because the struct fields are unknown
